libp2p/NodeTable.cpp

```diff
@@ -46,20 +46,18 @@ NodeTable::NodeTable(
     ba::io_service& _io, KeyPair const& _alias, NodeIPEndpoint const& _endpoint, bool _enabled)
   : m_hostNode(Node(_alias.pub(), _endpoint)),
     m_secret(_alias.secret()),
-    m_socket(make_shared<NodeSocket>(
-        _io, *reinterpret_cast<UDPSocketEvents*>(this), (bi::udp::endpoint)m_hostNode.endpoint)),
-    m_socketPointer(m_socket.get()),
+    m_socket(make_shared<NodeSocket>(_io, *this, (bi::udp::endpoint)m_hostNode.endpoint)),
```

Comment:
I would prefer to keep the inheritance from `UDPSocketEvents` private.
Then bacause of `make_shared` you would still need a cast here, but `static_cast` will do.
Also `m_socket` actually can be made `unique_ptr` I think, and then there'll be no `make_shared` and no cast here if you just use `new`.

Reply:
Comment to `m_socket` mentions that `shared_ptr` is required by boost.asio, but I don't believe this is true. The only sense in which I see it could be "required" is `shared_ptr` could have been used to avoid async handlers accessing already deleted variables.
But nowhere we capture this `shared_ptr` by value, and anyway `p2p::Host` takes care (at least tries) to stop IO service and all handlers before destroying `NodeTable` and other stuff.

Reply:
Unfortunately the comment is true. ASIO takes a weak_ptr of it somehow.

Reply:
Why do you want to have `UDPSocketEvents` private?

Reply:
Because deriving it is an implementation detail; the users of `NodeTable` don't need to know about it.